components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,242 @@ package xid
+	events, err := c.store.Get(ctx, since)
 	if err != nil {
 		return nil, err
 	}
-
-	ev, ok := last.Output.(*nvidia_query_nvml.XidEvent)
-	if !ok { // shoild never happen
-		return nil, fmt.Errorf("invalid output type: %T, expected nvidia_query_nvml.XidEvent", last.Output)
+	for _, event := range events {
+		ret = append(ret, resolveXIDEvent(event))
 	}
-	if ev != nil && ev.Xid > 0 {
-		o.NVMLXidEvent = ev
+	return ret, nil
+}
 
-		if lerr := c.poller.LastError(); lerr != nil {
-			log.Logger.Warnw("last query failed -- returning cached, possibly stale data", "error", lerr)
-		}
+func (c *XIDComponent) Metrics(ctx context.Context, since time.Time) ([]components.Metric, error) {
+	log.Logger.Debugw("querying metrics", "since", since)
 
-		lastSuccessPollElapsed := time.Now().UTC().Sub(ev.Time.Time)
-		if lastSuccessPollElapsed > 2*c.poller.Config().Interval.Duration {
-			log.Logger.Warnw("last poll is too old", "elapsed", lastSuccessPollElapsed, "interval", c.poller.Config().Interval.Duration)
-		}
-	}
+	return nil, nil
+}
 
-	return o, nil
+func (c *XIDComponent) Close() error {
+	log.Logger.Debugw("closing XIDComponent")
+	// safe to call stop multiple times
+	c.poller.Stop(nvidia_component_error_xid_id.Name)
+	c.cancel()
+	return nil
 }
 
-func (c *component) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
-	o, err := c.tailScan()
-	if err != nil {
-		return nil, err
+func (c *XIDComponent) start(watcher pkg_dmesg.Watcher) {
+	ticker := time.NewTicker(DefaultStateUpdatePeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.rootCtx.Done():
+			return
+		case <-ticker.C:
+			if err := c.updateCurrentState(); err != nil {
+				log.Logger.Errorw("failed to fetch current events", "error", err)
+				continue
+			}
+		case newEvent := <-c.extraEventCh:
+			if newEvent == nil {
+				continue
+			}
+			if err := c.store.Insert(c.rootCtx, *newEvent); err != nil {
+				log.Logger.Errorw("failed to create event", "error", err)
+				continue
+			}
+			events, err := c.store.Get(c.rootCtx, time.Time{})
+			if err != nil {
+				log.Logger.Errorw("failed to get all events", "error", err)
+				continue
+			}
+			c.mu.Lock()
+			c.currState = EvolveHealthyState(events)
+			c.mu.Unlock()
+		case dmesgLine := <-watcher.Watch():
+			log.Logger.Debugw("dmesg line", "line", dmesgLine)
+			ev, err := nvidia_query_xid.ParseDmesgLogLine(metav1.Time{Time: dmesgLine.Timestamp}, dmesgLine.Content)
+			if err != nil {
+				log.Logger.Errorw("failed to parse dmesg line", "error", err)
+				continue
+			}
+			if ev.Detail == nil {
+				log.Logger.Debugw("not xid event, skip")
+				continue
+			}
+			event := components.Event{
+				Time: metav1.Time{Time: dmesgLine.Timestamp},
+				Name: EventNameErroXid,
+				ExtraInfo: map[string]string{
+					EventKeyErroXidData: strconv.FormatInt(int64(ev.Detail.Xid), 10),
+					EventKeyDeviceUUID:  ev.DeviceUUID,
+				},
+			}
+			currEvent, err := c.store.Find(c.rootCtx, event)
```

Comment:
Not going to work... Find -> json.Marshal(event struct) -> json marshal .extraInfo -> emits unordered bytes (e.g., {a:b, 1:2} vs. {1:2, a:b}, and the select query will fail.

Let me see if we can make this more deterministic

Reply:
oh good catch..lemme fix

Reply:
See https://github.com/leptonai/gpud/pull/334 for fix.

Reply:
actually prob no need to make it ordered, can just remove the extraInfo query and go through extraInfo in findEvent